Simplify parser errors creation

nicolo-ribaudo opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` <!-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->

I love fp, but here it's just making the code much harder to follow.

nicolo-ribaudo commented on packages/babel-parser/src/parse-error.ts:
```diff
@@ -57,44 +74,39 @@ function toParseErrorConstructor<ErrorDetails extends object>({
+        return message;
+      },
+      set(value: string) {
+        Object.defineProperty(this, "message", { value, writable: true });
```
This was missing `writable: true`, causing the error in https://github.com/babel/babel/pull/9509.

liuxingbaoyu commented:
Is there something wrong with the `ParseErrorCode` enumeration? Personally I find enums more readable.

nicolo-ribaudo commented:
@liuxingbaoyu Moving it to `parser-error.ts` caused a circular dependency that that would break. I added it back as a `type` instead, so that we still get nice auto-completions and proper type-checking. I also added a bunch of `satisfies` checks to the error messages definitions, so that they are type-checked in-file rather than having errors reported in https://github.com/babel/babel/blob/b804bb08912d262133213dcd7f40d81645f97489/packages/babel-parser/src/parse-error.ts#L212-L217

liuxingbaoyu reviewed (approved):
Thanks! This looks great!
